src/main/kotlin/store/cloudtolocal/ImportData.kt

```diff
@@ -0,0 +1,68 @@
+package store.cloudtolocal
+
+import store.AppConfig
+import store.RealConfig
+import java.util.*
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+
+fun main() {
+    Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate({
+        runCatching {
+            RealConfig.importAllData()
+        }.onFailure {
+            println("🚨 Error in job")
+            it.printStackTrace()
+        }
+    }, 0, 1, TimeUnit.HOURS)
+}
+
+private fun AppConfig.importAllData() {
+    println("🏬 ------ Stores import job starting ${Date()} -------")
+    println("Importing stores info…")
+    importStoresInfo()
+    println("Importing extra fields…")
+    importExtraFields()
+    println("Importing seasons…")
+    importSeasons()
+    println("✅ ${Date()}")
+}
+
+private fun AppConfig.importStoresInfo() {
+    var page = 1
+    do {
+        print(" $page")
+        val result = runCatching {
+            storesProvider.listStores(page)
+        }.onSuccess {
+            it.map(storesRepository::saveInfo)
```

Comment:
To sum it up, it's very slow.
- If the migration is huge it can overlap with the next one
- If there's more stuff to do that depends on this data, it has to wait till it's imported since it can't be parallelized
- Lots of database calls are being made versus just one, which can affect database performance
However, it's important to remember that DB bulk operations may be dangerous because they skip business logic (e.g. conversions, validations, etc) and we might end up with invalid data in the DB.

Reply:
> However, it's important to remember that DB bulk operations may be dangerous because they skip business logic (e.g. conversions, validations, etc) and we might end up with invalid data in the DB.

I think this depends on the implementation. IMO inserting to the database (bulk or not) should be the very last step in that chain, i.e. after all conversions, validations, etc.

Also, if the database is properly designed and aligned with the business needs, it will be difficult to insert invalid or inconsistent data into it.

Reply:
you're saying to insert (bulk or not) through the middleware right?
apart from validations and conversions, imagine if the insertion of a store needs to send an email to that store. or it needs to generate events. or it needs auditing. etc. I wouldn't hardcode any business rule into the database anyway. "Databases are details" like uncle Bob says :)

Reply:
Going from bulk insert to hardcoding business rules into the database is a bit of a stretch.

The database model needs to be aligned with the business (eg. "stores can only have one owner", thus the database model should enforce this).

Regarding the operation being bulk or not, I don't see how it negatively affects the rest of the business logic.

* Stores get imported
* Each store gets validated/sanitized
* Invalid stores are discarded
* Generate any events prior to persisting
* Bulk insert (or a loop single insert, it shouldn't matter)
* If you want, an e-mail can be sent with a list of each new store (or any event)

Bulk or not, I think it boils down to performance. If your persistence logic is properly decoupled from the remaining of the project, it should have no effect on the business logic.

Reply:
I think we're saying the same thing then :)
Bulk insert through your domain rather than back-door insertion.
I think it's much harder (there might be operations you can't undo if the process fails) but feasible.